text/0000-deletable-immutable-data/0000-deletable-immutable-data.md

```diff
@@ -0,0 +1,68 @@
+
+Changing what is mentioned as a conceptual storage of `ImmutableData` in a node above can help achive this:
+```rust
+enum Owner {
+    Users(Vec<sign::PublicKey>),
+    Network,
+}
+
+struct ImmutableDataStorage {
+    immut_data: ImmutableData,
+    owner: Owner,
+}
+
+// New way of storage:
+let new_storage: HashMap<XorName, ImmutableDataStorage>;
+```
+Each time someone creates the same `ImmutableData`, one's `PublicKey` get appended to the `Owner::Users` field. Similarly each delete request from a user will be considered valid if the user's key is present in the list and if so the found key will be removed from the list. Once the length of `Owner::Users` reduces to zero, the data is deleted from the network. This field must allow duplication because one user can create same `ImmutableData` multiple times. For e.g. if `A` creates same `ImmutableData` thrice, `A` can issue 3 deletes all of which will be successful and will yeild a reimbursement. The last delete will actually delete the data from the network.
+
+There is however a downside - for popular data which might be deduplicated a large number of times, `Owner::Users` will grow enormously in size. A solution to this is that deletion of `ImmutableData` must be entail a _best-effort_ guarantee. If the deduplication rises above a limit (hence highly popular), we will consider that data to be entirely network owned and can no longer be deleted by any of its creators. At this point the storage will flip `owner` to `Owner::Network` and beyond this, `ImmutableData` can no longer be deleted and has no associated owners. The policy of reimbursement will not be applicable anymore.
+
+This upper limit shall be called `DEDUPLICATION_LIMIT` and shall be defined as follows: _The number of unique keys in `Owner::Users` beyond which this `ImmutableData` is entirely owned by the network_.
+Its value is chosen thus:
+```rust
+const DEDUPLICATION_LIMIT: usize = 20;
+```
+Once number of unique keys in `Owner::Users` is greater than `DEDUPLICATION_LIMIT`, an attempt to delete such an `ImmutableData` shall result in the following (existing) error:
+```rust
+// Defined in safe_network_common at the time of this RFC
+MutationError::InvalidOperation
+```
+
+## Drawbacks
+- A slight increase in complexity at vault level for handling of `ImmutableData`.
+- The deletion (and hence the incentive of reimbursement) is not a full guarantee but a best-effort one.
```

Comment:
I think another drawback may be churn. these lists will require to merge in cases where users are being added/removed from the list. During churn this will need to accumulate all the not yet locked lists.